torch/_subclasses/fake_tensor.py

```diff
@@ -392,6 +392,28 @@ def local_scalar_dense(fake_mode, func, arg):
         raise NotImplementedError(f"local_scalar_dense/item NYI for {arg.dtype}")
 
 
+@register_op_impl(lambda func: func is torch.ops.aten.nonzero.default)
+def nonzero(fake_mode, func, arg):
+    if fake_mode.shape_env is None:
+        # Without symints/symfloats, cannot handle this
+        raise DataDependentOutputException(func)
+    nnz = fake_mode.shape_env.create_unbacked_symint()
+    fake_mode.shape_env.expr_subs[nnz.node.expr].append(((nnz >= 0).node.expr, True))
+    fake_mode.shape_env.expr_subs[nnz.node.expr].append(((nnz < 0).node.expr, False))
+    fake_mode.shape_env.expr_subs[nnz.node.expr].append(((nnz == -1).node.expr, False))
+    # TODO: the dumb substitution here may not work well
+    fake_mode.shape_env.expr_subs[nnz.node.expr].append(((nnz < arg.numel()).node.expr, False))
+    fake_mode.shape_env.expr_subs[nnz.node.expr].append(((arg.numel() * nnz == 0).node.expr, False))
+    # TODO: the substitutions below are unsound
+    fake_mode.shape_env.expr_subs[nnz.node.expr].append(((nnz == 0).node.expr, False))
+    fake_mode.shape_env.expr_subs[nnz.node.expr].append(((nnz == 1).node.expr, False))
+    fake_mode.shape_env.expr_subs[nnz.node.expr].append(((nnz * arg.dim() == 0).node.expr, False))
+    # TODO: these are just hacks for the example
+    fake_mode.shape_env.expr_subs[nnz.node.expr].append(((nnz < 2).node.expr, False))
+    fake_mode.shape_env.expr_subs[nnz.node.expr].append(((9216 * nnz).node.expr, False))
+    return arg.new_empty((nnz, arg.dim()), dtype=torch.int64)
```

Comment:
I like, kind of suspect maybe the unbacked symint allocation should happen in meta tensor, not fake tensor. Main problem is getting our grubby hands on the ambient shapenv. Maybe torch._guards is the ticket.